File: LinkedList/SimpleLinkedList/SimpleLinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class SimpleLinkedList:
    def __init__(self):
        self.head = None

    def prepend(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node

    def append(self, value):
        new_node = Node(value)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    def insert(self, value, position):
        if position == 0:
            self.prepend(value)
            return
        new_node = Node(value)
        current = self.head
        for _ in range(position - 1):
            if current is None:
                raise IndexError("Position out of bounds")
            current = current.next
        if current is None:
            raise IndexError("Position out of bounds")
        new_node.next = current.next
        current.next = new_node

    def deleteFirst(self):
        if not self.head:
            return
        self.head = self.head.next

    def deleteLast(self):
        if not self.head:
            return
        if not self.head.next:
            self.head = None
            return
        current = self.head
        while current.next.next:
            current = current.next
        current.next = None

    def delete(self, position):
        if not self.head:
            return
        if position == 0:
            self.deleteFirst()
            return
        current = self.head
        for _ in range(position - 1):
            if current.next is None:
                return
            current = current.next
        if current.next is None:
            return
        current.next = current.next.next

    def search(self, value):
        current = self.head
        position = 0
        while current:
            if current.value == value:
                return position
            current = current.next
            position += 1
        return -1

    def print(self):
        current = self.head
        elements = []
        while current:
            elements.append(str(current.value))
            current = current.next
        print(" -> ".join(elements) if elements else "Empty List")

if __name__ == "__main__":
    ll = SimpleLinkedList()
    ll.append(10)
    ll.append(20)
    ll.prepend(5)
    ll.insert(15, 2)
    ll.print()
    print(f"Position of 15: {ll.search(15)}")
    ll.delete(2)
    ll.print()

File: LinkedList/SimpleLinkedList/SimpleLinkedList.java
public class SimpleLinkedList {
    private static class Node {
        int value;
        Node next;
        Node(int value) {
            this.value = value;
            this.next = null;
        }
    }

    private Node head = null;

    public void prepend(int value) {
        Node newNode = new Node(value);
        newNode.next = head;
        head = newNode;
    }

    public void append(int value) {
        Node newNode = new Node(value);
        if (head == null) {
            head = newNode;
            return;
        }
        Node current = head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = newNode;
    }

    public void insert(int value, int position) {
        if (position == 0) {
            prepend(value);
            return;
        }
        Node newNode = new Node(value);
        Node current = head;
        for (int i = 0; i < position - 1; i++) {
            if (current == null) return;
            current = current.next;
        }
        if (current == null) return;
        newNode.next = current.next;
        current.next = newNode;
    }

    public void deleteFirst() {
        if (head == null) return;
        head = head.next;
    }

    public void deleteLast() {
        if (head == null) return;
        if (head.next == null) {
            head = null;
            return;
        }
        Node current = head;
        while (current.next.next != null) {
            current = current.next;
        }
        current.next = null;
    }

    public void delete(int position) {
        if (head == null) return;
        if (position == 0) {
            deleteFirst();
            return;
        }
        Node current = head;
        for (int i = 0; i < position - 1; i++) {
            if (current.next == null) return;
            current = current.next;
        }
        if (current.next == null) return;
        current.next = current.next.next;
    }

    public int search(int value) {
        Node current = head;
        int position = 0;
        while (current != null) {
            if (current.value == value) return position;
            current = current->next; 
            current = current.next;
            position++;
        }
        return -1;
    }

    public void print() {
        Node current = head;
        if (current == null) {
            System.out.println("Empty List");
            return;
        }
        while (current != null) {
            System.out.print(current.value);
            if (current.next != null) System.out.print(" -> ");
            current = current.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        SimpleLinkedList ll = new SimpleLinkedList();
        ll.append(10);
        ll.append(20);
        ll.prepend(5);
        ll.insert(15, 2);
        ll.print();
    }
}
